### PR TITLE
Add Backward Compatibility Testing documentation

### PR DESCRIPTION
**Pull request by mpanchajanya**
### What this PR does / why we need it

Add Tanzu Plugin Runtime Backward Compatibility Testing documentation.

### CODE REVIEW

File: test/BACKWARD_COMPATIBILITY.md [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+Run all `****_test.go` files from `compatibility-tests` directory.
+
+``` shell
+make backward-compatibility-tests
+```
+
+### How are the Tests being run ?
+
+The tests are executed as a GitHub runner CI pipeline that executes the below command.
```
Comment (vuil) [inline test/BACKWARD_COMPATIBILITY.md:25]: This section is not particularly useful (and can be pretty much assumed). I was hoping for some information on what to look for in terms of what tests are run, and how the results are presented. Please take a note and we can iterate on this later.

File: test/BACKWARD_COMPATIBILITY.md [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+}
+
+type DeleteContextOutputOptions struct{
+ error string // expected error message could be the sub string of actual error message
+}
+```
+
+### xxxOptions structs
+
+- These structs are super set of parameters for respective configs like Context, Server, DiscoverySources etc. for all Runtime Versions.
```
Comment (vuil) [inline test/BACKWARD_COMPATIBILITY.md:124]:
```suggestion
- Each of these struct represents the super set of all fields for their respective config types (e.g .Context, Server, DiscoverySource) across for all Runtime versions in which the types are defined.
```

File: test/BACKWARD_COMPATIBILITY.md [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+- Based on the Runtime Version xxxOptions attributes may change(mandatory/optional/Not applicable).
+- It is meant to be used as both inputOptions and outputOptions.
+- Command Creation Helper functions (i.e. NewSetContextCommand) validate the supplied inputOptions as per RuntimeVersion and make sure all required attributes are set. If not Command Creation Helper functions will throw error and Test fails.
+- The validation of “Input and expect output” happens in  Set/Get/DeleteXXXCommand creation (within framework.NewTestCase), so it's not part of test case execution. so its kind of test case setup.
+- Input struct: Framework validates only Input data (eg: GetContextInputOptions/SetContextInputOptions/DeleteContextInputOptions) as its required to make runtime api call,
```
Comment (vuil) [inline test/BACKWARD_COMPATIBILITY.md:129]: Not sure what you mean here. Are you saying that framework validates only the parts of the input data that are required to make a particular version of the API call, and ignores any other information?

File: test/BACKWARD_COMPATIBILITY.md [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+- Output struct: Framework validates the supplied outputOptions as per specified runtime version whether certain fields are supported for the runtime version used and compares the supplied data with what is returned from API method i.e. every non-null field in output struct will be checked for equality with returned result.
+  (Eg: GetContextOutputOptions/SetContextOutputOptions/DeleteContextOutputOptions) as its test writer responsibility, the Output data is depends on the Input data (eg: if input is not valid, runtime api may return error) OR the sequence of calls made before the current call (eg: NewGetContextCommand() output depends on the previous SetContext** calls).
+  If test writer is not passing valid expected output data, then it fails during the TEST CASE EXECUTION as test case fail.
+- Test framework does validate the Output struct’s null and non-null values with runtime api response.
```
Comment (vuil) [inline test/BACKWARD_COMPATIBILITY.md:133]: I am not sure what this means. If the framework does not validate the output's null and non-null values against the API response, does it validate _anything_ about the output at all?

File: test/BACKWARD_COMPATIBILITY_FRAMEWORK.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
+# Backward Compatibility
+
+## Framework
+
+### How to write a test?
+
+Each Test is of below type
```
Comment (vuil) [inline test/BACKWARD_COMPATIBILITY_FRAMEWORK.md:7]: I am not sure what this line mean.

File: test/BACKWARD_COMPATIBILITY_FRAMEWORK.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+``` go
+// Call runtime GetContext() API function
+actualCtx := GetContext(ctxName) //i.e. test-bc
+
+// Print the log to stdout
+```
+
+### Combination / Co - Existence Testing
+
```
Comment (vuil) [inline test/BACKWARD_COMPATIBILITY_FRAMEWORK.md:191]: what does this table mean, it needs an explanation.

File: test/BACKWARD_COMPATIBILITY_FRAMEWORK.md (hunk 5 above)
Comment (vuil) [reply test/BACKWARD_COMPATIBILITY_FRAMEWORK.md:7]: Is BACKWARD_COMPATIBILITY_FRAMEWORK.md meant more for showing the inner workings of how the test case is processed by the framework to then invoke various APIs and perform validations?

If so, please provide an intro at the top of the doc, and outline what is going to be discussed. 
If we can, in a couple of sentences, just explain how the test 'cases' is being transformed into something than can be consumed by binaries build with specific versions of the runtime, the rest of the writeup might be easier to follow.

File: README.md [hunk 7]
```diff
@@ -14,3 +14,12 @@ Developers can use the `Builder` admin plugin to bootstrap a new plugin which ca
 Tanzu Plugin Runtime provides various config api methods to perform CRUD operations on Contexts, Servers, DiscoverySources, Features, Envs etc.
```
[the remark this answers is not in the document]
Comment (mpanchajanya) [reply README.md:14]: done

File: README.md [hunk 8]
```diff
@@ -14,3 +14,12 @@ Developers can use the `Builder` admin plugin to bootstrap a new plugin which ca
...
 
 For more details about the design and apis go to [Config API](docs/config.md)
+
+## Backward Compatibility Testing
+
+### Overview
+
+Backward Compatibility Runtime testing ensures that APIs of different Tanzu Plugin Runtime versions work as expected as long as those Tanzu Plugin Runtime versions are supported.
```
[the remark this answers is not in the document]
Comment (mpanchajanya) [reply README.md:22]: updated to point the APIs to Config APIs doc

File: README.md [hunk 9]
```diff
@@ -14,3 +14,12 @@ Developers can use the `Builder` admin plugin to bootstrap a new plugin which ca
...
+Backward Compatibility Runtime testing verifies that APIs with new Tanzu Plugin Runtime (Ex :- v1.0.0) works along with old APIs of Tanzu Plugin Runtime versions like v0.28.0, v0.25.4, v0.11.6.
```
[the remark this answers is not in the document]
Comment (mpanchajanya) [reply README.md:23]: Updated

File: test/BACKWARD_COMPATIBILITY.md [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+
+## Summary
+
+Compatibility Test Framework provides test helper functions to write compatibility test cases.
+
+Test writer is responsible for writing the go tests using the framework provided helper methods.
+As a test writer you should be knowledgeable on specific APIs and specific Runtime libraries i.e. what arguments each API takes and what it returns.
+Test writers should also be familiar with how to write combination tests with multiple APIs i.e. when testing mutators and reader APIs together.
+
+For Example:-  Running SetContext alone in the test will not be a good test as it does not validate the actual context that is set in the config yaml file. But when SetContext is combined with GetContext/ GetCurrentContext making the tests more solid.
```
[the remark this answers is not in the document]
Comment (mpanchajanya) [reply test/BACKWARD_COMPATIBILITY.md:11]: Yea make sense  updated to include generic statements

File: test/BACKWARD_COMPATIBILITY.md (hunk 2 above)
Comment (mpanchajanya) [reply test/BACKWARD_COMPATIBILITY.md:124]: updated

File: test/BACKWARD_COMPATIBILITY.md (hunk 3 above)
Comment (mpanchajanya) [reply test/BACKWARD_COMPATIBILITY.md:129]: Re phrased the sentence to `- Input struct: Framework validates the supplied Input data (eg: GetContextInputOptions/SetContextInputOptions/DeleteContextInputOptions) as per specified runtime version supported fields.
`

File: test/BACKWARD_COMPATIBILITY.md (hunk 4 above)
Comment (mpanchajanya) [reply test/BACKWARD_COMPATIBILITY.md:133]: Rephrased the statement to `- Test framework does validate the expected Output struct’s field values with runtime api response to its entirety.
`

File: test/BACKWARD_COMPATIBILITY_FRAMEWORK.md (hunk 6 above)
Comment (mpanchajanya) [reply test/BACKWARD_COMPATIBILITY_FRAMEWORK.md:191]: added summary

File: test/BACKWARD_COMPATIBILITY_FRAMEWORK.md (hunk 5 above)
Comment (mpanchajanya) [reply test/BACKWARD_COMPATIBILITY_FRAMEWORK.md:7]: updated the details

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+
+### What will be the test results ?
+
+GitHub CI runner pipeline include Logs with details on test cases from test-case that are executed and succeeded/failed.
+
+## API and Test cases details
+
+### Test Suite API
+
+Framework has `framework.NewTestCase` to write a specific test case, each test case consist sequence of commands, each command is specific to a runtime api call.
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:41]:
```suggestion
Framework has `framework.NewTestCase` to write a specific test case, each test case consists of a sequence of commands, with each command corresponding to an invocatino of a API from a specific version of runtime.
```

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+We need to use the above high level APIs to construct specific test command, the input data for each high level apis changes based on runtime version's its being tested, for more details look into the input parameter's type definition.
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:42]: high level APIs -> do you mean the test helper functions? The only API mentioned so far are plugin runtime APIs. If you intend to refer to the framework test APIs, perhaps indicate clearly like in L51.

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+
+``` go
+// Construct series of commands to execute
+ testCase := framework.NewTestCase().Add(setContextCommand).Add(getContextCommand)
+```
+
+### High-level test API
+
+To write compatibility test cases, the test writer should be aware of below high level framework Test APIs, each below api helps to test specific runtime api by constructing test command:
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:51]: there are still api/API inconsistencies are various parts of the docs

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+func NewSetXXXCommand(setXXXInputOptions SetXXXInputOptions, setXXXOutputOptions SetXXXOutputOptions) (*Command, error)
+func NewGetXXXCommand(getXXXInputOptions GetXXXInputOptions, getXXXOutputOptions GetXXXOutputOptions) (*Command, error)
+```
+
+``` go
+
+// NewSetContextCommand contructs a command to make a call to specific runtime version SetContext API.
+// Input Parameter: setContextInputOptions has all input parameters which are required for Runtime SetContext API.
+// Input Parameter: setContextOutputOptions has details about expected output from Runtime SetContext API call.
+// Return:  command to execute or error if any validations fails for SetContextInputOptions or SetContextOutputOptions
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:63]:
```suggestion
// Return: command to execute or error if any validations fails for SetContextInputOptions or SetContextOutputOptions
```

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+// This method does validates the input parameters  SetContextInputOptions/SetContextOutputOptions based on Runtime API Version
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:64]: missing a word or two here near the double space gap?

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+}
+
+type DeleteContextOutputOptions struct{
+ error string // expected error message could be the sub string of actual error message
+}
+```
+
+### xxxOptions structs
+
+- Each of these struct represents the super set of all fields for their respective config types (e.g .Context, Server, DiscoverySource) across for all Runtime versions in which the types are defined.
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:120]:
```suggestion
- Each of these structs represents the super set of all fields for their respective config types (e.g .Context, Server, DiscoverySource) across all Runtime versions in which the types are defined.
```

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+  If test writer is not passing valid expected output data, then it fails during the TEST CASE EXECUTION as test case fail.
+- Test framework does validate the expected Output struct’s field values with runtime api response to its entirety.
+- Test writer does not have to explicitly specify to check whether an attribute value is nil as all fields default values are nil or “”.
+- Test writer will need to specify what exact object attribute values he expects out of GetXX api method (framework perform a full equality check and not partial equality, means if api returns attribute which not specified by test writer then test case will fail).
+- So for the output struct, all filed are optional, there are NO MANDATORY fields!!! its test writer responsibility to identify the expected data!!
+
+Example 1: For the Below NewSetContextCommand if the context arguments are passed incorrect
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:134]: Please be consistent with casing of the words. Any reason why Below is capitalized.

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+Below Error is thrown when the test case with above command input options is run
+    Expected
+        <*errors.fundamental | 0x140000b8030>: {
+            msg: "invalid set context input options for the specified runtime version v0.28.0",
+            stack: [0x1051f2ac9, 0x1051f1bf0, 0x1051f4544, 0x10500a120, 0x105009c64, 0x105009474, 0x10500c478, 0x10500bec8, 0x1050285c0, 0x10502833c, 0x105027ad0, 0x105029b5c, 0x105034da0, 0x105034c44, 0x1051f4d38, 0x104e27fd0, 0x104d91814],
+        }
+    to be nil
+```
+
+Example 2  Explains on how framework validates the null fields and What is expected of Test Writer
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:161]:
```suggestion
Example 2  How framework validates the null fields and what is expected of test writer
```

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+
+type DeleteContextOutputOptions struct{
+ error string // expected error message could be the sub string of actual error message
+}
+```
+
+### xxxOptions structs
+
+- Each of these struct represents the super set of all fields for their respective config types (e.g .Context, Server, DiscoverySource) across for all Runtime versions in which the types are defined.
+- Based on the Runtime Version xxxOptions attributes may change(mandatory/optional/Not applicable).
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:121]:
```suggestion
- Based on the Runtime Version xxxOptions attributes may change (mandatory/optional/Not applicable).
```

File: test/CROSS_VERSION_API_COMPATIBILITY.md [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+- It is meant to be used as both inputOptions and outputOptions.
+- Command Creation Helper functions (i.e. NewSetContextCommand) validate the supplied inputOptions as per RuntimeVersion and make sure all required attributes are set. If not Command Creation Helper functions will throw error and Test fails.
+- The validation of “Input and expect output” happens in  Set/Get/DeleteXXXCommand creation (within framework.NewTestCase), so it's not part of test case execution. so its kind of test case setup.
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY.md:124]:
```suggestion
- The validation of “Input and expect output” happens in Set/Get/DeleteXXXCommand creation (within framework.NewTestCase), so it's not part of test case execution. so its kind of test case setup.
```

File: test/CROSS_VERSION_API_COMPATIBILITY.md (hunk 19 above)
Comment (vuil) [reply test/CROSS_VERSION_API_COMPATIBILITY.md:121]: combining L121-122:
The xxxOptions structs are used as inputs to the API invocation as well as data to verify against the output of the invocation.The implications of setting or not setting certain attributes changes depending which scenario under which it is being used.

File: test/CROSS_VERSION_API_COMPATIBILITY_FRAMEWORK.md [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+```
+
+``` go
+// Call runtime GetContext() API function
+actualCtx := GetContext(ctxName) //i.e. test-bc
+
+// Print the log to stdout
+```
+
+### Combination / Co - Existence Testing
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY_FRAMEWORK.md:192]:
```suggestion
### Combination / Co-existence Testing
```

File: test/CROSS_VERSION_API_COMPATIBILITY_FRAMEWORK.md [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+
+Below table describes all possible combination tests for all supported Runtime APIs.
```
Comment (vuil) [inline test/CROSS_VERSION_API_COMPATIBILITY_FRAMEWORK.md:194]: I do not understand the meaning of this sentence or the tables below.
Here, taking an example of one or two cells in the table and explain in english what it means will be useful.

**Review by vuil — COMMENTED** [review]
I think the doc is looking quite good now. Most comments are nits and easy to address. 
I requested more explanation of the last section with the tables.
Also use lower-case-hyphen-style.md for naming MD files.
Some of the toplevel files are uppercase due to GH conventions, but other files should follow this pattern.

**Review by vuil — APPROVED** [review]

**Comment by vuil** [comment]
Feel free to merge after address this set of comments

File: test/CROSS_VERSION_API_COMPATIBILITY.md (hunk 12 above)
Comment (mpanchajanya) [reply test/CROSS_VERSION_API_COMPATIBILITY.md:42]: updated and moved to High level test API section

File: test/CROSS_VERSION_API_COMPATIBILITY.md (hunk 13 above)
Comment (mpanchajanya) [reply test/CROSS_VERSION_API_COMPATIBILITY.md:51]: addressed all now

File: test/CROSS_VERSION_API_COMPATIBILITY.md (hunk 15 above)
Comment (mpanchajanya) [reply test/CROSS_VERSION_API_COMPATIBILITY.md:64]: fixed

File: test/CROSS_VERSION_API_COMPATIBILITY.md (hunk 19 above)
Comment (mpanchajanya) [reply test/CROSS_VERSION_API_COMPATIBILITY.md:121]: updated

File: test/CROSS_VERSION_API_COMPATIBILITY_FRAMEWORK.md (hunk 22 above)
Comment (mpanchajanya) [reply test/CROSS_VERSION_API_COMPATIBILITY_FRAMEWORK.md]: This section is not relevant anymore. removed. moved to phase 2 doc

**Pull request merged**